docs/testing.md

```diff
@@ -0,0 +1,212 @@
+      defaultStore,
+    );
+
+    await waitFor(() => {
+      expect(queryByText(/Very likely/u)).toBeInTheDocument();
+    });
+  });
+
+  it('renders unlikely estimate', async () => {
+    const props = generateGasEstimatorProps({
+      maxPriorityFeePerGas: '100',
+    });
+
+    const { queryByText } = renderWithProvider(
+      <GasEstimator {...props} />,
+      defaultStore,
+    );
+
+    await waitFor(() => {
+      expect(queryByText(/Very unlikely/u)).toBeInTheDocument();
+    });
+  });
+});
+```
+
+Using this pattern, each test contains the minimum sufficient information for setting up the assertion (setting the `maxPriorityFeePerGas`).
+
+This also means test data is not entirely reused between different tests. In each test, the relevant data is explicitly stated.
+
+The data generation functions can be placed at the top of the file if they are used exclusively in it or shared files if they are reused for testing multiple components or units of the repository.
+
+### Using Jest timers
+
+Imperative waiting for time to pass in a test locks the test execution more than it needs to and can cause flakiness and timeouts on slower machines and so it should generally be avoided.
+
+Here's an example of that:
+
+```javascript
+// Wait 5 seconds
+await new Promise<void>((resolve) => setTimeout(() => resolve(), 5000));
+```
+
+Jest has a [set of timer APIs](https://jestjs.io/docs/timer-mocks) to replace that anti-pattern. From the docs:
+
+> The native timer functions (i.e., setTimeout(), setInterval(), clearTimeout(), clearInterval()) are less than ideal for a testing environment since they depend on real-time to elapse. Jest can swap out timers with functions that allow you to control the passage of time.
+
+Using `jest.advanceTimersByTime(msToRun)`, we can tell Jest what effect we want to simulate on the test (declarative programming) instead of having to provoke the effect (imperative programming) using the promisified `setTimeout`.
+
+Unfortunately, using these APIs With asynchronous code leaves unresolved promises, so we can use a custom `flushPromises` method to resolve them and let Jest optimize the execution speed in the background ([source](https://stackoverflow.com/questions/52177631/jest-timer-and-promise-dont-work-well-settimeout-and-async-function/58716087#58716087)):
+
+```javascript
+function flushPromises(): Promise<unknown> {
+  return new Promise(jest.requireActual('timers').setImmediate);
+}
+
+async function advanceTimersByTime(ms) {
+  jest.advanceTimersByTime(ms);
+  await flushPromises();
+}
+```
+
+Here's [an example PR](https://github.com/MetaMask/core/pull/1002/files) for this pattern in our core library.
+
+For UI tests, React Testing Library exposes another API that relies on the same idea of abstracting the passage of time instead of explicitly managing it:
+
+```javascript
+waitFor(() => {
+  expect(<element that needs will disappear once the DOM updates>).toBeRemoved()
+})
+```
+
```

Comment:
It might also be beneficial to include
- when should we add snapshot test
- how to test redux

wdyt？

Reply:
see comment above https://github.com/MetaMask/metamask-extension/pull/17411#discussion_r1101158884